### DISCUSSION TITLE
Multi Tenant strategy help - turtles all the way down

### DISCUSSION THREAD

**Discussion by corepay** [zitadel/zitadel#7800] [Q&A]
I could use some help or ideas here on how to accomplish a multi-tenant strategy that goes several levels deep.  This is what I've come up with so far and it can result in several thousand organizations  if I take this path...is there a better way?

* I have a several TENANT organizations that are top-level private label affairs.
* Each tenant provides merchant account services that ride on my payment switch.  A tenant can can have one, several or several thousand merchant accounts that have one or more users that log into my merchant portal
* Each merchant account can have several or several thousand customers that can have one or more users login to the merchant ecommerce account, customer portal and/or the merchnat markeplace.

I'm having trouble reasoning through where organizations should stop and some other group strategy maybe by using role groups should begin if at all.  

All account and user management will occur in my apps interfacing with zitadel through apis - no one should or have access to zitadel.

The simplest way I **think** I can get this done is just create organizations all the way down.  For example:

1.  Me, the system admin creates tenant accounts from my top-level org.  When I create the tenant account I also GRANT access to a PROJECT in my organization that has the APPS the tenant needs to create and manage merchants.  One of the PROJECTS for example is a tenant portal, another an underwriting app, another PROJECT is a CRM of sorts and another PROJECT a merchant on-boarding module...

2.  When I create a tenant I can GRANT access to projects that I want the tenant to access - for example the TENANT may pay for the underwriting app but does not want the CRM app so I do not GRANT the CRM app.   

When I board a TENANT I also create PROJECTS  the tenant organization that the merchants will need to use.  For example one of the PROJECTS will be a merchant portal they can log into, another a CUSTOMER portal, another a VENDOR portal, another a POS/Virtual termianl, another a Billing engine, another an ecommerce site, etc.  Same strategy here, when a tenant boards a merchant they will be able to GRANT specific projects to a merchant that are premium or paid for or however they want to do it.  For example, the TENANT may provide the Marketplace app option for $49 a month.  If the merchant signs up for that the tenant will grant that Project App to the merchant.

>  Now the tenant has the GRANT to use PROJECT apps from my top-level SYSTEM org required to manage portfolios of merchants like underwriting and onboarding and reports and the TENANT org also has a bunch of PROJECT apps they can GRANT to their customer - the merchants.

3.  When a TENANT boards a new merchant, I'll code it so the MERCHANT gets PROJECTS and APPS that the TENANT selects from some checkbox or something.  For example if the TENANT selects "Customer Portal" when boarding the merchant I'll create a new ORG for the merchant, GRANT the default PROJECT app 'merchant portal' (basic features) to the new merchant ORG and also create a new PROJECT in the new merchant ORG that contains the Customer Portal app. 

> Now the merchant ORG has access to a merchant portal, can manage their own users and can GRANT access to the PROJECT Customer portal app to their customers.

4.  Now we are at the merchant ORG level.  When the merchant creates a customer I will create new customer ORG with acccess to the merchant ORG PROJECT with the customer app.  The customer will be able to create other users for their account but not be able to create any orgs.

Each step of the way a designated admin account for each ORG will be the owner of the created ORG <- I think this will link these all up to maintain my heirachy and relationships.

So my burning questions right now are:
1.  Is this feasable?  Is it over-engineered or delusional or a management nightmare or just not doable?
2.  I can have a dozen or more TENANTS, each tenant can have 10k merchants and each merchant can have hundreds of customers.  I understand I can scale Zitadel out multiple load-balanced instances and can over-engineer my dbs, so hardware aside can Zitadel handle 100k organizations?  If it is a namespace I don't see this being too big a deal but I just don't know enough yet.
3.  yes this is a lot of work - I'm thinking through a simple ROLE and group strategy that can help organize some of this and I'm good at building loops so Im not concerned too much about the overhead of creating these ORGS with code / apis...but am I delusional at this scale?
4.   Is there a better way?  For example, maybe instead of the merchants creating customer organizations I can come up with some role/group strategy that isolated each customer account - but I'm looking ahead to the marketplace and merchant B2B customers and vendors so the ability for customers to create their own user accounts and roles would be good.
5.  At the end of the day, I assume that me, at the top/top level will have visibility and ability to access and manage all the way down the chain by some combination of authorization, invites and impersonation. Sí?
6.  The last thing I can think of I'm not sure about is if there is a way to link organizations without having a user of the parent organization be an 'owner' of the child orgs.  I would prefer to have the parent be the ORG and not the User and the user I create when I create the child org be the owner.  I see I can create an ORG owner when creatng an org by uncheking "use your personal acount when creating this org' option....but then I assume if I do that I lose  the parent/child relationship.

Well that was a lot - if you made it this far trust me your brain doesn't hurt as much as mine righ tnow...nonetheless I would really appreciate a little feedback if anything jumped out or maybe a link to something I haven;t stumbled across in the docs.

Much appreciated!!

**Comment by hifabienne** [comment]
Hei @corepay 
Wow, thats a lot.
I have to admit I do not understand everything fully, but as far from what I understand it makes sense.
Maybe some inputs that might help you.

1. Our general recommendation is to create organizations when you have different people administrating the users. 
Example:
- You own org is managed by you
- You merchant manages its own users

This makes it easier to also give self management rights to managers of an organizations so you do not have to do everything by yourself.

2. We do not yet have a group possibility, the groups you can see on the roles are more for a ux topic, so you can tell the ui to select all roles from a specific group. In the token you will only get the role keys. of course you can make something like group1:admin in the role keys. We do have an issue for groups, which is not yet implemented: https://github.com/zitadel/zitadel/issues/5822

3. Organizations are not really in a hirarchy. But with the grants you can achieve what you need. Just be aware, that it is only possible in one level. So you can grant a project from org a to org b, but you can't grant the project from org a to org b to org c

I hope this input does help you
